File: Dart/solutions/day07.dart
import '../utils/index.dart';

class Day07 extends GenericDay {
  Day07() : super(7);

  @override
  List<CamelCard> parseInput() {
    return input.getPerLine().map((e) {
      final [hand, bid, ...] = e.split(' ');
      return CamelCard(hand: hand, bid: int.parse(bid));
    }).toList();
  }

  @override
  int solvePart1() {
    final sortedCards = parseInput()..sort();
    print(sortedCards.take(10));
    return sortedCards.foldIndexed<int>(
      0,
      (index, previous, element) => previous += (index + 1) * element.bid,
    );
  }

  @override
  int solvePart2() {
    return 0;
  }
}

class CamelCard implements Comparable<CamelCard> {
  const CamelCard({
    required this.hand,
    required this.bid,
  }) : assert(
          hand.length == 5,
          'A hand in Camel Card game always has 5 cards',
        );

  final String hand;
  final int bid;

  @override
  String toString() {
    return 'CamelCard(hand: $hand, bid: $bid)';
  }

  HandType getHandType() {
    return HandType.highCard;
  }

  HandOfCards get handOfCards {
    final [first, second, third, fourth, fifth, ...] = hand.split('');
    return (
      first: CardLabel.fromString(first),
      second: CardLabel.fromString(second),
      third: CardLabel.fromString(third),
      fourth: CardLabel.fromString(fourth),
      fifth: CardLabel.fromString(fifth),
    );
  }

  List<CardLabel> get cardsList => handOfCards.toList();

  HandType get handType {
    switch (handOfCards.toSet().length) {
      case 1:
        return HandType.fiveOfKind;
      case 4:
        return HandType.onePair;
      case 5:
        return HandType.highCard;
      case 2:
        final newCards = [...cardsList]
          ..removeWhere((element) => element == handOfCards.first);
        if (newCards.length == 2 || newCards.length == 3) {
          return HandType.fullHouse;
        }
        return HandType.fourOfKind;
      case 3:
        for (var i = 0; i < 5; i++) {
          final newCards = [...cardsList]
            ..retainWhere((element) => element == cardsList[i]);
          if (newCards.length > 1) {
            if (newCards.length == 2) {
              return HandType.twoPair;
            }
            return HandType.threeOfKind;
          }
        }
        throw StateError('Impossible State');

      default:
        throw StateError('Invalid Hand');
    }
  }

  @override
  int compareTo(CamelCard other) {
    final handTypeComparison = handType.compareTo(other.handType);
    if (handTypeComparison == 0) {
      final otherCards = other.cardsList;
      for (final (index, card) in cardsList.indexed) {
        if (card.compareTo(otherCards[index]) != 0) {
          return card.compareTo(otherCards[index]);
        }
      }
    }
    return handTypeComparison;
  }
}

enum HandType implements Comparable<HandType> {
  fiveOfKind(6),
  fourOfKind(5),
  fullHouse(4),
  threeOfKind(3),
  twoPair(2),
  onePair(1),
  highCard(0);

  const HandType(this.value);
  final int value;

  @override
  int compareTo(HandType other) {
    return value.compareTo(other.value);
  }
}

typedef HandOfCards = ({
  CardLabel first,
  CardLabel second,
  CardLabel third,
  CardLabel fourth,
  CardLabel fifth
});

extension on HandOfCards {
  Set<CardLabel> toSet() {
    return <CardLabel>{first, second, third, fourth, fifth};
  }

  List<CardLabel> toList() {
    return <CardLabel>[first, second, third, fourth, fifth];
  }
}

enum CardLabel implements Comparable<CardLabel> {
  ace(13),
  king(12),
  queen(11),
  jack(10),
  ten(9),
  nine(8),
  eight(7),
  seven(6),
  six(5),
  five(4),
  four(3),
  three(2),
  two(1);

  const CardLabel(this.value);
  final int value;

  @override
  int compareTo(CardLabel other) {
    return value.compareTo(other.value);
  }

  static CardLabel fromString(String label) {
    return switch (label) {
      'A' => CardLabel.ace,
      'K' => CardLabel.king,
      'Q' => CardLabel.queen,
      'J' => CardLabel.jack,
      'T' => CardLabel.ten,
      '9' => CardLabel.nine,
      '8' => CardLabel.eight,
      '7' => CardLabel.seven,
      '6' => CardLabel.six,
      '5' => CardLabel.five,
      '4' => CardLabel.four,
      '3' => CardLabel.three,
      '2' => CardLabel.two,
      _ => throw StateError('Invalid Label'),
    };
  }
}

File: Swift/Sources/Day07.swift
import Algorithms

struct Day07: AdventDay {
  // Save your data in a corresponding text file in the `Data` directory.
  let data: String

  private let entities: [(hand: Hand, bid: Int)]

  private init(data: String, entities: [(hand: Hand, bid: Int)]) {
    self.data = data
    self.entities = entities
  }

  init(data: String) {
    self.init(
      data: data,
      entities:
        data.split(separator: "\n").map {
          let handBid = $0.split(separator: " ")
          return (Hand(using: String(handBid[0])), Int(handBid[1])!)
        }
    )
  }

  private func calculateWinnings(usingRule rule: CamelCardRule) -> Int {
    let sorted = entities.sorted(by: {
      lhs, rhs in
      let left = lhs.hand
      let right = rhs.hand
      return left.weakerThan(other: right, usingRule: rule)
    })
    return sorted.map(\.bid).enumerated().reduce(0) {
      value, enumeratedBid in
      let rank = enumeratedBid.offset + 1
      return enumeratedBid.element * rank + value
    }
  }

  func part1() -> Int {
    calculateWinnings(usingRule: .normal)
  }

  func part2() -> Int {
    calculateWinnings(usingRule: .jokers)
  }
}

private enum Card: Character {
  case ace = "A"
  case king = "K"
  case queen = "Q"
  case jack = "J"
  case ten = "T"
  case nine = "9"
  case eight = "8"
  case seven = "7"
  case six = "6"
  case five = "5"
  case four = "4"
  case three = "3"
  case two = "2"

  var points: Int {
    return switch self {
    case .ace:
      14
    case .king:
      13
    case .queen:
      12
    case .jack:
      11
    case .ten:
      10
    case .nine:
      9
    case .eight:
      8
    case .seven:
      7
    case .six:
      6
    case .five:
      5
    case .four:
      4
    case .three:
      3
    case .two:
      2
    }
  }
}

private enum HandType: Int, Comparable {

  case highCard = 1
  case onePair, twoPair, threeOfKind, fullHouse, fourOfKind, fiveOfKind

  static func < (lhs: HandType, rhs: HandType) -> Bool {
    lhs.rawValue < rhs.rawValue
  }
}

private enum CamelCardRule {
  case normal
  case jokers
}

private struct Hand {
  let cards: [Card]
  let cardCounts: [Card: Int]

  init(using hand: String) {
    self.cards = Array(hand).compactMap { Card(rawValue: $0) }
    self.cardCounts = cards.reduce([Card: Int]()) {
      value, card in
      var newValue = value
      newValue[card, default: 0] += 1
      return newValue
    }
  }

  func weakerThan(other: Hand, usingRule rule: CamelCardRule) -> Bool {
    if handType(usingRule: rule) == other.handType(usingRule: rule) {
      if let (index, card) = cards.enumerated().first(where: { index, card in
        card != other.cards[index]
      }) {
        switch rule {
        case .normal:
          return card.points < other.cards[index].points
        case .jokers:
          let myPoints =
            if card == .jack {
              0
            } else { card.points }
          let otherPoints =
            if other.cards[index] == .jack {
              0
            } else { other.cards[index].points }
          return myPoints < otherPoints
        }
      }
    }
    return handType(usingRule: rule) < other.handType(usingRule: rule)
  }

  func handType(usingRule rule: CamelCardRule = .normal) -> HandType {
    assert(cards.count == 5, "A Hand in Camel Cards has exactly 5 cards")
    if rule == .jokers && !cardCounts.keys.contains(Card.jack) {
      return handType()
    }

    let distinctCards = cardCounts.count
    switch rule {
    case .normal:
      return switch distinctCards {
      case 1:
        .fiveOfKind
      case 2:
        if cardCounts.values.contains(4) {
          .fourOfKind
        } else {
          .fullHouse
        }
      case 3:
        if cardCounts.values.contains(3) {
          .threeOfKind
        } else {
          .twoPair
        }
      case 4:
        .onePair
      default:
        .highCard
      }

    case .jokers:
      // If this case is encountered then we always have at least one Card.jack i.e jockerCount > 0
      let jokerCount = cardCounts[.jack, default: 0]
      switch distinctCards {
      case 1:
        return .fiveOfKind
      case 2:
        return .fiveOfKind
      case 3:
        if jokerCount == 1 {
          if cardCounts.values.contains(2) {
            return .fullHouse
          }
        }
        return .fourOfKind
      case 4:
        return .threeOfKind
      default:
        return .onePair
      }
    }
  }
}
